pkg/ipam/crd.go

```diff
@@ -233,6 +233,31 @@ func deriveVpcCIDR(node *ciliumv2.CiliumNode) (result *cidr.CIDR) {
 	return
 }
 
+func (n *nodeStore) autoDetectIPv4NativeRoutingCIDR() {
+	if vpcCIDR := deriveVpcCIDR(n.ownNode); vpcCIDR != nil {
+		if nativeCIDR := n.conf.GetIPv4NativeRoutingCIDR(); nativeCIDR != nil {
+			logFields := logrus.Fields{
+				"vpc-cidr":                   vpcCIDR.String(),
+				option.IPv4NativeRoutingCIDR: nativeCIDR.String(),
+			}
+
+			ranges4, _ := ip.CoalesceCIDRs([]*net.IPNet{nativeCIDR.IPNet, vpcCIDR.IPNet})
+			if len(ranges4) != 1 {
+				log.WithFields(logFields).Fatal("Native routing CIDR does not contain VPC CIDR.")
+			} else {
+				log.WithFields(logFields).Info("Ignoring autodetected VPC CIDR.")
+			}
+		} else {
+			log.WithFields(logrus.Fields{
+				"vpc-cidr": vpcCIDR.String(),
+			}).Info("Using autodetected VPC CIDR.")
+			n.conf.SetIPv4NativeRoutingCIDR(vpcCIDR)
+		}
+	} else {
+		log.Fatal("Could not determine VPC CIDR")
```

Comment:
I'm not sure I fully understand why this is a non-recoverable case. Yes, `node` is created by the agent (i.e. the node itself), but `node.Status.ENI.ENIs` is populated by the operator at a later time. 

Meaning that it can very well be that the initial version of the `CiliumNode` CRD doesn't contain the ENI information yet and we have to wait till the operator has populated the ENI info we're waiting on here.

Reply:
@gandro I did not take this into account, my assumption was that a missing VPC cidr means that the VPN could not be queried and thus we are in an unrecoverable state. I changed the first commit to replicate the old logic.